modules/multisignatures.js

```diff
@@ -67,171 +74,254 @@ function Multisignatures(cb, scope) {
-					}
-					multisignatureTransaction.signatures =
-						multisignatureTransaction.signatures || [];
-					multisignatureTransaction.signatures.push(transaction.signature);
-					multisignatureTransaction.ready = Multisignature.prototype.ready(
-						multisignatureTransaction,
-						sender
-					);
-
-					library.bus.message('signature', transaction, true);
-					return setImmediate(cb);
-				}
-			);
-		}, cb);
-	}
 
-	if (!multisignatureTransaction) {
-		return setImmediate(cb, 'Transaction not found');
-	}
+	// Add signature to transaction
+	transaction.signatures.push(signature.signature);
+	// Check if transaction is ready to be processed
+	transaction.ready = library.logic.multisignature.ready(transaction, sender);
 
-	if (multisignatureTransaction.type === transactionTypes.MULTI) {
-		multisignatureTransaction.signatures =
-			multisignatureTransaction.signatures || [];
+	// Emit events
+	library.network.io.sockets.emit(
+		'multisignatures/signature/change',
+		transaction
+	);
+	library.bus.message('signature', signature, true);
+	return setImmediate(cb);
+};
 
-		if (
-			multisignatureTransaction.asset.multisignature.signatures ||
-			multisignatureTransaction.signatures.indexOf(transaction.signature) !== -1
-		) {
-			return setImmediate(cb, 'Permission to sign transaction denied');
-		}
+/**
+ * Process signature for multisignature account creation, transaction type 4 (MULTI)
+ *
+ * @private
+ * @param {Object} signature - Signature data
+ * @param {string} [signature.publicKey] - Public key of account that created the signature (optional)
+ * @param {string} signature.transactionId - Id of transaction that signature was created for
+ * @param {string} signature.signature - Actual signature
+ * @param {Object} transaction - Corresponding transaction grabbed from transaction pool
+ * @param {function} cb - Callback function
+ * @implements {__private.validateSignature}
+ * @returns {setImmediateCallback} cb, err
+ */
+__private.processSignatureForMultisignatureAccountCreation = (
+	signature,
+	transaction,
+	cb
+) => {
+	// Normalize members of multisignature account from transaction
+	const membersPublicKeys = transaction.asset.multisignature.keysgroup.map(
+		member => member.substring(1) // Remove first character, which is '+'
+	);
 
-		// Find public key
-		let verify = false;
-
-		try {
-			for (
-				let i = 0;
-				i < multisignatureTransaction.asset.multisignature.keysgroup.length &&
-				!verify;
-				i++
-			) {
-				const key = multisignatureTransaction.asset.multisignature.keysgroup[
-					i
-				].substring(1);
-				verify = library.logic.transaction.verifySignature(
-					multisignatureTransaction,
-					key,
-					transaction.signature
-				);
-			}
-		} catch (e) {
-			library.logger.error(e.stack);
-			return setImmediate(cb, 'Failed to verify signature');
-		}
+	// Set empty object as sender, as we don't need sender in case of multisignature registration
+	const sender = {};
```

Comment:
Are we sure that we don't need to get sender account? What would happen if there multiple multisig registration transactions from the same account?

Reply:
I checked logic in `Multisignature.prototype.ready` and passing `sender` in that case is not needed because it will never contains `multisignatures` property (account is not yet multisignature account). There is a non-explicit check for exactly that thing - when transaction is multisignature registration then compare things from `transaction` object itself instead of `sender`.